File: adapters/python/examples/modbus_tcp/adapter.py
"""Modbus TCP adapter that plugs into BaseAdapter.

Reads each register spec independently (one round-trip per spec). This
keeps the example linear — production adapters with hundreds of tags
should batch reads by contiguous address range; that optimization is
out of scope for this reference.
"""

from __future__ import annotations

import logging

from pymodbus.client import AsyncModbusTcpClient
from pymodbus.exceptions import (
    ConnectionException,
    ModbusException,
    ModbusIOException,
)

from sdk import BaseAdapter, DeviceConnectionError, DeviceTimeoutError, TagValue

from .config import ModbusConfig
from .decoder import RegisterSpec, decode_register

logger = logging.getLogger(__name__)


class ModbusTCPAdapter(BaseAdapter):
    """Reference Modbus TCP adapter.

    Subclass or copy this for real deployments. The configuration is
    expected to come from a YAML mapping file; see config.load_config().
    """

    def __init__(self, config: ModbusConfig, **kwargs) -> None:
        super().__init__(**kwargs)
        self._cfg = config
        self._client: AsyncModbusTcpClient | None = None

    async def connect_device(self) -> None:
        self._client = AsyncModbusTcpClient(
            host=self._cfg.host,
            port=self._cfg.port,
            timeout=self._cfg.timeout,
        )
        ok = await self._client.connect()
        if not ok or not self._client.connected:
            raise DeviceConnectionError(
                f"failed to connect to modbus://{self._cfg.host}:{self._cfg.port}"
            )

    async def disconnect_device(self) -> None:
        if self._client is not None:
            self._client.close()
            self._client = None

    async def check_device_health(self) -> None:
        # Use the first register as a cheap ping. If the device dropped,
        # this read will fail and the SDK reconnect path takes over.
        if self._client is None or not self._client.connected:
            raise DeviceConnectionError("modbus client not connected")
        first = self._cfg.registers[0]
        try:
            await self._read_words(first)
        except (DeviceConnectionError, DeviceTimeoutError):
            raise
        except ModbusException as e:
            raise DeviceConnectionError(f"health check failed: {e}") from e

    async def collect(self) -> list[TagValue]:
        values: list[TagValue] = []
        for spec in self._cfg.registers:
            words = await self._read_words(spec)
            values.append(decode_register(words, spec))
        return values

    async def _read_words(self, spec: RegisterSpec) -> list[int]:
        if self._client is None:
            raise DeviceConnectionError("modbus client not initialized")

        count = spec.word_count
        try:
            if spec.function == "holding":
                resp = await self._client.read_holding_registers(
                    address=spec.address,
                    count=count,
                    device_id=self._cfg.unit_id,
                )
            elif spec.function == "input":
                resp = await self._client.read_input_registers(
                    address=spec.address,
                    count=count,
                    device_id=self._cfg.unit_id,
                )
            else:
                raise DeviceConnectionError(
                    f"unsupported function: {spec.function}"
                )
        except ConnectionException as e:
            raise DeviceConnectionError(str(e)) from e
        except ModbusIOException as e:
            raise DeviceTimeoutError(str(e)) from e

        if resp.isError():
            # Modbus exception responses (e.g. illegal address) propagate
            # here. We surface as DeviceConnectionError so the SDK's
            # backoff loop kicks in instead of crashing the adapter.
            raise DeviceConnectionError(
                f"modbus error for register {spec.name!r}: {resp}"
            )
        return list(resp.registers)

File: adapters/go/sdk/examples/modbus_tcp_sensor/adapter.go
package main

import (
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/goburrow/modbus"

	"github.com/e7217/edg/adapters/go/sdk"
)

// ModbusDevice implements sdk.Collector + sdk.DeviceLifecycle for a single
// Modbus TCP slave. One ModbusDevice serves all registers from one host.
type ModbusDevice struct {
	cfg *ModbusConfig

	mu      sync.Mutex
	handler *modbus.TCPClientHandler
	client  modbus.Client
}

// NewModbusDevice builds a device but does not connect; the SDK calls
// ConnectDevice when the adapter starts.
func NewModbusDevice(cfg *ModbusConfig) *ModbusDevice {
	return &ModbusDevice{cfg: cfg}
}

func (d *ModbusDevice) ConnectDevice(_ context.Context) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	h := modbus.NewTCPClientHandler(fmt.Sprintf("%s:%d", d.cfg.Host, d.cfg.Port))
	h.Timeout = time.Duration(d.cfg.Timeout * float64(time.Second))
	h.SlaveId = d.cfg.UnitID
	if err := h.Connect(); err != nil {
		return fmt.Errorf("%w: %v", sdk.ErrDeviceConnection, err)
	}
	d.handler = h
	d.client = modbus.NewClient(h)
	return nil
}

func (d *ModbusDevice) DisconnectDevice(_ context.Context) error {
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.handler != nil {
		_ = d.handler.Close()
		d.handler = nil
		d.client = nil
	}
	return nil
}

func (d *ModbusDevice) CheckDeviceHealth(ctx context.Context) error {
	// Cheap liveness check: read the first configured register.
	d.mu.Lock()
	client := d.client
	d.mu.Unlock()
	if client == nil {
		return fmt.Errorf("%w: not connected", sdk.ErrDeviceConnection)
	}
	if len(d.cfg.Registers) == 0 {
		return nil
	}
	_, err := d.readWords(d.cfg.Registers[0])
	return err
}

func (d *ModbusDevice) Collect(_ context.Context) ([]sdk.TagValue, error) {
	d.mu.Lock()
	client := d.client
	d.mu.Unlock()
	if client == nil {
		return nil, fmt.Errorf("%w: not connected", sdk.ErrDeviceConnection)
	}

	values := make([]sdk.TagValue, 0, len(d.cfg.Registers))
	for _, spec := range d.cfg.Registers {
		words, err := d.readWords(spec)
		if err != nil {
			return nil, err
		}
		tv, err := DecodeRegister(words, spec)
		if err != nil {
			return nil, err
		}
		values = append(values, tv)
	}
	return values, nil
}

func (d *ModbusDevice) readWords(spec RegisterSpec) ([]uint16, error) {
	count, err := spec.WordCount()
	if err != nil {
		return nil, err
	}

	d.mu.Lock()
	client := d.client
	d.mu.Unlock()
	if client == nil {
		return nil, fmt.Errorf("%w: not connected", sdk.ErrDeviceConnection)
	}

	var raw []byte
	switch spec.Function {
	case "holding":
		raw, err = client.ReadHoldingRegisters(spec.Address, count)
	case "input":
		raw, err = client.ReadInputRegisters(spec.Address, count)
	default:
		return nil, fmt.Errorf("unsupported function: %s", spec.Function)
	}
	if err != nil {
		// goburrow surfaces Modbus exception responses and transport
		// errors both via the returned error. We don't try to
		// distinguish here — both feed the SDK's reconnect loop.
		return nil, fmt.Errorf("%w: read %s @%d: %v",
			sdk.ErrDeviceConnection, spec.Function, spec.Address, err)
	}

	words, err := bytesToWordsErr(raw)
	if err != nil {
		return nil, err
	}
	return words, nil
}

// bytesToWords converts goburrow's raw register payload (2 bytes per
// register, big-endian) into a slice of uint16 register values.
func bytesToWords(b []byte) []uint16 {
	w, _ := bytesToWordsErr(b)
	return w
}

func bytesToWordsErr(b []byte) ([]uint16, error) {
	if len(b)%2 != 0 {
		return nil, errors.New("packet length is not a multiple of 2")
	}
	out := make([]uint16, len(b)/2)
	for i := range out {
		out[i] = binary.BigEndian.Uint16(b[i*2:])
	}
	return out, nil
}
